src/support/qsort.c

```diff
@@ -0,0 +1,269 @@
+/*-
+ * Copyright (c) 1992, 1993
+ *    The Regents of the University of California.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "wt_internal.h"
+
+#define WT_QSORT_INSERT_THRESHOLD 7
+#define WT_QSORT_LARGE_THRESHOLD 40
+
+#ifdef WT_QSORT_R_DEF
+typedef int wt_cmp_t(const void *, const void *, void *);
+#else
+typedef int wt_cmp_t(const void *, const void *);
+#endif
+
+#ifdef WT_QSORT_R_DEF
+#define CMP(x, y, t) (cmp((x), (y), (t)))
+#else
+#define CMP(x, y, t) (cmp((x), (y)))
+#endif
+
+/*
+ * __swap_bytes --
+ *     Swap the contents of two arbitrary values of a given size.
+ */
+static inline void
+__swap_bytes(uint8_t *a, uint8_t *b, size_t size)
```

Comment:
Consider using the `restrict` keyword on pointers `a` and `b`. Rationale: it makes no sense for them to be aliasing.

Reply:
Thanks, good call - done in dfd9a2f.

Reply:
MSVC doesn't like this, had to revert in bdd1457.